src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -1335,19 +1335,25 @@ private Symbol UnwrapAlias(Symbol symbol, out AliasSymbol alias, BindingDiagnost
         private ImmutableArray<TypeWithAnnotations> BindTypeArguments(SeparatedSyntaxList<TypeSyntax> typeArguments, BindingDiagnosticBag diagnostics, ConsList<TypeSymbol> basesBeingResolved = null)
         {
             Debug.Assert(typeArguments.Count > 0);
-            var args = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+            var args = ArrayBuilder<TypeWithAnnotations>.GetInstance(typeArguments.Count);
             foreach (var argSyntax in typeArguments)
-            {
                 args.Add(BindTypeArgument(argSyntax, diagnostics, basesBeingResolved));
-            }
 
             return args.ToImmutableAndFree();
         }
 
         private TypeWithAnnotations BindTypeArgument(TypeSyntax typeArgument, BindingDiagnosticBag diagnostics, ConsList<TypeSymbol> basesBeingResolved = null)
         {
-            // Unsafe types can never be type arguments, but there's a special error code for that.
-            var binder = this.WithAdditionalFlags(BinderFlags.SuppressUnsafeDiagnostics);
+            // BackCompat.  The compiler would previously suppress reporting errors for pointers in generic types.  This
+            // was intended so you would get a specific error in CheckBasicConstraints.CheckBasicConstraints for a type
+            // like (like `List<int*>`). i.e. you would get the error about an unsafe type not being a legal type argument,
+            // but not the error about not being in an unsafe context.  This had the unfortunate consequence though of 
+            // preventing the latter check for something like `List<int*[]>`.  Here, this is a legal generic type, but we 
+            // still want to report the error that you need to be in an unsafe context.  So, to maintain compat, we only
+            // do the suppression if you're on C# 11 and prior.  In later versions we do the correct check.
+            var binder = ((CSharpParseOptions)typeArgument.SyntaxTree.Options).LanguageVersion < LanguageVersion.Preview
```

Comment:
>LanguageVersion.Preview [](http://example.com/codeflow?start=97&length=23)

I do not think we want to hardcode ```LanguageVersion.Preview``` here. Should probably request the language version for ```MessageID.IDS_FeatureUsingTypeAlias```.

Reply:
wfm.